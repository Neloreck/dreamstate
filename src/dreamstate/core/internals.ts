import type { Context } from "react";

import type {
  TAnyContextManagerConstructor,
  TAnyValue,
  TQuerySubscriptionMetadata,
  TSignalSubscriptionMetadata,
} from "@/dreamstate/types";

/**
 * Meta symbols for private internals in context managers.
 */
export const SIGNAL_METADATA_SYMBOL: unique symbol = Symbol("SIGNAL_METADATA");
export const QUERY_METADATA_SYMBOL: unique symbol = Symbol("QUERY_METADATA");
export const SIGNALING_HANDLER_SYMBOL: unique symbol = Symbol("SIGNALING_HANDLER");
export const SCOPE_SYMBOL: unique symbol = Symbol("SCOPE");

/**
 * A weak map registry that stores React context instances bound to specific manager classes.
 *
 * This registry ensures that the library does not retain unnecessary references to manager classes,
 * preventing memory leaks and unintended side effects.
 *
 * This is particularly useful in scenarios such as:
 * - Hot Module Replacement
 * - Module unloading
 * - Scope disposal
 */
export const CONTEXT_REACT_CONTEXTS_REGISTRY: WeakMap<
  TAnyContextManagerConstructor,
  Context<TAnyValue>
> = new WeakMap();

export const SIGNAL_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TSignalSubscriptionMetadata> =
  new WeakMap();

export const QUERY_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TQuerySubscriptionMetadata> =
  new WeakMap();
