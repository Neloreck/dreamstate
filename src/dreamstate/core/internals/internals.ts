import type { Context } from "react";

import type {
  TServiceMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber,
  TDreamstateService,
  TAnyObject
} from "@/dreamstate/types";

/**
 * Global shared constants.
 */

export const EMPTY_ARR: [] = [];

/**
 * Store related constants.
 */

// Currently active services instance types.
export const CONTEXT_SERVICES_ACTIVATED: Set<TDreamstateService> = new Set();

// Currently active services instance types.
export const CONTEXT_SERVICES_REFERENCES: TServiceMap<number> = new WeakMap();

// ReactContext registry, lazy initialized constants.
export const CONTEXT_REACT_CONTEXTS_REGISTRY: TServiceMap<Context<any>> = new WeakMap();

// ContextService instances registry.
export const CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TDreamstateService>> = new WeakMap();

// ContextManagers context references registry.
export const CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject> = new WeakMap();

// ContextManagers observers registry - Providers of related context.
export const CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>> = new WeakMap();

// ContextManagers subscribers registry for data updates.
export const CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>> = new WeakMap();

// Signal handlers registry bound to context instances.
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TServiceMap<TSignalListener<any>> = new WeakMap();

// Signal listeners metadata registry related to services.
export const CONTEXT_SIGNAL_METADATA_REGISTRY: TServiceMap<TSignalSubscriptionMetadata> = new WeakMap();

// Query listeners metadata registry related to services.
export const CONTEXT_QUERY_METADATA_REGISTRY: TServiceMap<TQuerySubscriptionMetadata> = new WeakMap();

// Global signal listeners registry.
export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();
