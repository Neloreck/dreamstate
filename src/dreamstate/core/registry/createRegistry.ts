import {
  TAnyContextManagerConstructor, TAnyObject,
  TQueryListener,
  TQueryType,
  TServiceMap,
  TSignalListener, TUpdateObserver, TUpdateSubscriber
} from "@/dreamstate/types";

export interface IRegistry {
  SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>>;
  QUERY_PROVIDERS_REGISTRY: Map<TQueryType, Array<TQueryListener<any, any>>>;
  CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TAnyContextManagerConstructor>>;
  CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject>;
  CONTEXT_SERVICES_REFERENCES: TServiceMap<number>;
  CONTEXT_SERVICES_ACTIVATED: Set<TAnyContextManagerConstructor>;
  // ContextManagers observers registry - Providers of related context.
  CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>>;
  // ContextManagers subscribers registry for data updates.
  CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>>;
}

/**
 * Create registry with internal stores.
 */
export function createRegistry(): IRegistry {
  return ({
    CONTEXT_SERVICES_ACTIVATED: new Set(),
    QUERY_PROVIDERS_REGISTRY: new Map(),
    SIGNAL_LISTENERS_REGISTRY: new Set(),
    CONTEXT_SERVICES_REGISTRY: new Map(),
    CONTEXT_STATES_REGISTRY: new Map(),
    CONTEXT_SERVICES_REFERENCES: new Map(),
    CONTEXT_OBSERVERS_REGISTRY: new Map(),
    CONTEXT_SUBSCRIBERS_REGISTRY: new Map()
  });
}
