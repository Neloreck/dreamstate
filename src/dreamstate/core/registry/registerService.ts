import {
  CONTEXT_SERVICES_REGISTRY,
  CONTEXT_OBSERVERS_REGISTRY,
  CONTEXT_SIGNAL_HANDLERS_REGISTRY,
  CONTEXT_STATES_REGISTRY,
  CONTEXT_SUBSCRIBERS_REGISTRY,
  CONTEXT_SERVICES_ACTIVATED,
  CONTEXT_SERVICES_REFERENCES
} from "@/dreamstate/core/internals";
import { processComputed } from "@/dreamstate/core/observing/processComputed";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { onMetadataSignalListenerCalled } from "@/dreamstate/core/signals/onMetadataSignalListenerCalled";
import { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
import { TAnyObject, TDreamstateService, TSignalListener } from "@/dreamstate/types";

/**
 * Register context manager entry.
 */
export function registerService<S extends any = any>(
  Service: TDreamstateService<any>,
  initialState?: S
): void {
  // Only if registry is empty -> create new instance, remember its context and save it to registry.
  if (!CONTEXT_SERVICES_REGISTRY.has(Service)) {
    const instance: InstanceType<TDreamstateService<any>> = new Service(initialState);
    const signalHandler: TSignalListener = onMetadataSignalListenerCalled.bind(instance);

    CONTEXT_OBSERVERS_REGISTRY.set(Service, CONTEXT_OBSERVERS_REGISTRY.get(Service) || new Set());
    CONTEXT_SERVICES_REGISTRY.set(Service, instance);
    CONTEXT_SIGNAL_HANDLERS_REGISTRY.set(Service, signalHandler);
    CONTEXT_SERVICES_REFERENCES.set(Service, 0);

    subscribeToSignals(signalHandler);

    CONTEXT_SERVICES_ACTIVATED.add(Service);

    // Currently only context managers require additional information initialization.
    if (Service.prototype instanceof ContextManager) {
      CONTEXT_STATES_REGISTRY.set(Service, (instance as ContextManager<TAnyObject>).context);
      // Subscribers are not always sync, should not block their un-sub later.
      CONTEXT_SUBSCRIBERS_REGISTRY.set(
        Service,
        CONTEXT_SUBSCRIBERS_REGISTRY.get(Service) || new Set()
      );

      // todo: Add checkContext method call for deb bundle with warnings for initial state nesting.
      processComputed((instance as ContextManager<any>).context);
    }
  }
}
