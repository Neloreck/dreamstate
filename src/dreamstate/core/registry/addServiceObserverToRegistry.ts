import { CONTEXT_SERVICES_REGISTRY, CONTEXT_OBSERVERS_REGISTRY } from "@/dreamstate/core/internals";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import { TDreamstateService, TUpdateObserver } from "@/dreamstate/types";

/**
 * Add state changes observer.
 */
export function addServiceObserverToRegistry(
  Service: TDreamstateService,
  observer: TUpdateObserver
): void {
  CONTEXT_OBSERVERS_REGISTRY.get(Service)!.add(observer);

  // Notify about provision, if it is first observer.
  if (CONTEXT_OBSERVERS_REGISTRY.get(Service)!.size === 1) {
    (CONTEXT_SERVICES_REGISTRY.get(Service) as ContextService)["onProvisionStarted"]();
  }
}
