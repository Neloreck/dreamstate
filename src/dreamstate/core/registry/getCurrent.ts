import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyObject, TDreamstateService } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TDreamstateService<any>>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
