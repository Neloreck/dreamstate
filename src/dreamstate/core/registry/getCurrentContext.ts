import { debug } from "@/macroses/debug.macro";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current supplied context.
 */
export function getCurrentContext<T extends TAnyContextManagerConstructor>(
  Manager: T
): T["prototype"]["context"] | null {
  debug.info("Requested current manager context:", Manager.name);

  return CONTEXT_STATES_REGISTRY.get(Manager) || null;
}
