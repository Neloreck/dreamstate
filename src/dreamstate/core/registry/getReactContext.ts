import { Context, createContext } from "react";

import { debug } from "@/macroses/debug.macro";

import { CONTEXT_REACT_CONTEXTS_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get manager react context internal.
 */
export function getReactContext<T extends TAnyContextManagerConstructor>(
  Manager: T
): Context<T> {
  debug.info("Requested manager react context:", Manager.name);

  if (CONTEXT_REACT_CONTEXTS_REGISTRY.has(Manager)) {
    return CONTEXT_REACT_CONTEXTS_REGISTRY.get(Manager)!;
  } else {
    const reactContext: Context<T> = createContext(null as any);

    reactContext.displayName = "DS." + Manager.name;

    CONTEXT_REACT_CONTEXTS_REGISTRY.set(Manager, reactContext);

    debug.info("Context manager context declared:", Manager.name, reactContext.displayName);

    return reactContext;
  }
}
