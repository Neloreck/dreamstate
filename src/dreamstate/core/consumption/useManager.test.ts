import { mount, render } from "enzyme";
import { createElement, ReactElement, useEffect } from "react";
import { act } from "react-dom/test-utils";

import { createProvider, useManager } from "@/dreamstate";
import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { getCurrent } from "@/dreamstate/test-utils/registry/getCurrent";
import { registerService } from "@/dreamstate/test-utils/registry/registerService";
import { nextAsyncQueue } from "@/dreamstate/test-utils/utils/nextAsyncQueue";
import { TAnyContextManagerConstructor, TAnyObject, TCallable } from "@/dreamstate/types";
import {
  TestContextManager,
  ITestContext
} from "@/fixtures";

describe("UseManager subscription and rendering", () => {
  function FunctionalConsumerWithUseEffect<T extends TAnyContextManagerConstructor>({
    onUpdate,
    onCheckContextDiff
  }: {
    onUpdate: TCallable;
    onCheckContextDiff?: (context: T["prototype"]["context"]) => Array<any>;
  }): ReactElement {
    const context: TAnyObject = useManager(TestContextManager as any, onCheckContextDiff as any);

    useEffect(onUpdate);

    return createElement("div", {}, JSON.stringify(context));
  }

  function FunctionalConsumer(): ReactElement {
    const value: ITestContext = useManager(TestContextManager);

    return createElement("span", {}, JSON.stringify(value));
  }

  function FunctionalConsumerWithMemo(): ReactElement {
    const value: ITestContext = useManager(TestContextManager, ({ second }) => [ second ]);

    return createElement("span", {}, JSON.stringify(value));
  }

  it("Functional components should properly subscribe to managers without diff-checking cb", async () => {
    const getRoot = () => (
      createElement(
        createProvider([ TestContextManager ]),
        {},
        createElement(FunctionalConsumerWithUseEffect, {
          onUpdate: mockFn
        })
      )
    );

    const mockFn = jest.fn();
    const tree = mount(getRoot());

    expect(mockFn).toHaveBeenCalledTimes(1);
    expect(render(getRoot())).toMatchSnapshot();

    mockFn.mockClear();

    const manager: TestContextManager = getCurrent(TestContextManager)!;

    await act(async () => {
      manager.setContext({ second: manager.context.second });
      await nextAsyncQueue();
    });

    expect(mockFn).not.toHaveBeenCalled();

    mockFn.mockClear();

    await act(async () => {
      manager.setContext({ second: 100 });
      await nextAsyncQueue();
    });

    expect(mockFn).toHaveBeenCalled();

    mockFn.mockClear();

    await act(async () => {
      manager.setContext({ first: "anotherString" });
      await nextAsyncQueue();
    });

    expect(mockFn).toHaveBeenCalled();

    mockFn.mockClear();
    tree.unmount();
  });

  it("Functional components should properly subscribe to managers with diff-checking cb", async () => {
    const getRoot = () => (
      createElement(
        createProvider([ TestContextManager ]),
        {},
        createElement(FunctionalConsumerWithUseEffect, {
          onUpdate: mockFn,
          onCheckContextDiff: (context: ITestContext) => [ context.first ]
        })
      )
    );

    const mockFn = jest.fn(() => {});
    const tree = mount(getRoot());

    expect(render(getRoot())).toMatchSnapshot();
    expect(mockFn).toHaveBeenCalledTimes(1);

    mockFn.mockClear();

    const manager: TestContextManager = getCurrent(TestContextManager)!;

    await act(async () => {
      manager.setContext({ second: 100 });
      await nextAsyncQueue();
    });

    expect(mockFn).not.toHaveBeenCalled();
    mockFn.mockClear();

    await act(async () => {
      manager.setContext({ first: "anotherString" });
      await nextAsyncQueue();
    });

    expect(mockFn).toHaveBeenCalled();

    mockFn.mockClear();
    tree.unmount();
  });

  it("Should properly fire provision events for functional observers", async () => {
    const manager: TestContextManager = registerService(TestContextManager);

    manager["onProvisionStarted"] = jest.fn();
    manager["onProvisionEnded"] = jest.fn();

    const tree = mount(
      createElement(
        createProvider([ TestContextManager ]),
        {},
        createElement(FunctionalConsumer, {})
      )
    );

    await nextAsyncQueue();

    expect(manager["onProvisionStarted"]).toHaveBeenCalled();

    tree.unmount();

    await nextAsyncQueue();
    await nextAsyncQueue();

    expect(manager["onProvisionEnded"]).toHaveBeenCalled();
    expect(CONTEXT_SERVICES_REGISTRY.has(TestContextManager)).toBeFalsy();
  });

  it("Should properly fire provision events for functional subscribers", async () => {
    const manager: TestContextManager = registerService(TestContextManager);

    manager["onProvisionStarted"] = jest.fn();
    manager["onProvisionEnded"] = jest.fn();

    const tree = mount(
      createElement(
        createProvider([ TestContextManager ]),
        {},
        createElement(FunctionalConsumerWithMemo, {})
      )
    );

    await nextAsyncQueue();

    expect(manager["onProvisionStarted"]).toHaveBeenCalled();

    tree.unmount();

    await nextAsyncQueue();

    expect(manager["onProvisionEnded"]).toHaveBeenCalled();
    expect(CONTEXT_SERVICES_REGISTRY.has(TestContextManager)).toBeFalsy();
  });
});
