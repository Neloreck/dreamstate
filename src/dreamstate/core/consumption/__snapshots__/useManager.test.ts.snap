// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UseManager subscription and rendering. Functional components should properly subscribe to managers with diff-checking cb. 1`] = `
<Memo(DS.Observer)>
  <FunctionalConsumerWithUseEffect
    onCheckContextDiff={[Function]}
    onUpdate={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <div>
      {"first":"first","second":2,"third":false}
    </div>
  </FunctionalConsumerWithUseEffect>
</Memo(DS.Observer)>
`;

exports[`UseManager subscription and rendering. Functional components should properly subscribe to managers without diff-checking cb. 1`] = `
<Memo(DS.Observer)>
  <FunctionalConsumerWithUseEffect
    onUpdate={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <div>
      {"first":"first","second":2,"third":false}
    </div>
  </FunctionalConsumerWithUseEffect>
</Memo(DS.Observer)>
`;
